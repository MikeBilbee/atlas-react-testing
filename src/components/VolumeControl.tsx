import React, { useState, useEffect } from 'react'; // Import useEffect
import { SpeakerXMarkIcon, SpeakerWaveIcon } from '@heroicons/react/24/solid';

interface VolumeControlProps {
	volume: number;
	onChange: (newVolume: number) => void;
}

const VolumeControl: React.FC<VolumeControlProps> = ({ volume: initialVolume, onChange }) => { // Destructure and rename prop
	const [volume, setVolume] = useState(initialVolume);

	useEffect(() => {
		setVolume(initialVolume);
	}, [initialVolume]);

	const handleVolumeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const newVolume = parseInt(event.target.value, 10);
		setVolume(newVolume);

		onChange(newVolume);
	};

	return (
		<div className="flex items-center space-x-2">
			{volume === 0 ? (
				<SpeakerXMarkIcon data-testid="mute-icon" className="h-5 w-5 text-gray-300 dark:text-gray-800" />
			) : (
				<SpeakerWaveIcon data-testid="speaker-icon" className="h-5 w-5 text-gray-600 dark:text-gray-400" />
			)}

		<div className={`relative min-w-64 sm:w-35 md:w-45 lg:w-96 h-2 bg-gray-200 dark:bg-gray-400 rounded-full`}>
			<div
				className={`absolute h-full bg-light-background dark:bg-dark-secondary rounded-full transition-all duration-300 ease-in-out`}
				style={{ width: `${volume}%` }}
			/>
				<input
					type="range"
					min="0"
					max="100"
					value={volume}
					onChange={handleVolumeChange}
					className={`
						absolute w-full h-full appearance-none bg-transparent focus:outline-none
						&::-webkit-slider-thumb {
							-webkit-appearance: none;
							appearance: none;
							width: 12px;
							height: 12px;
							background-color: #4b5563;
							border-radius: 50%;
							cursor: pointer;
						}
					`}
				/>
			</div>
		</div>
	);
};

export default VolumeControl;