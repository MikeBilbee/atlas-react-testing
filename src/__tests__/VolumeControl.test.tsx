import { describe, it, expect, vi } from 'vitest';
import { render, fireEvent } from '@testing-library/react';
import VolumeControl from '../components/VolumeControl'; 

describe('VolumeControl component', () => {

  it('renders correctly with initial volume', () => {
    const { container } = render(<VolumeControl volume={50} onChange={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  it('calls onChange prop when volume changes', () => {
    const onChangeMock = vi.fn();
    const { container } = render(<VolumeControl volume={30} onChange={onChangeMock} />);

    const slider = container.querySelector('input[type="range"]') as HTMLInputElement;

    fireEvent.change(slider, { target: { value: '75' } });
    expect(onChangeMock).toHaveBeenCalledWith(75);
  });

  it('updates the progress bar width on volume change', () => {
    const { container } = render(<VolumeControl volume={30} onChange={() => {}} />);

    const slider = container.querySelector('input[type="range"]') as HTMLInputElement;
    const progressBar = container.querySelector('.bg-light-background') as HTMLElement; 

    fireEvent.change(slider, { target: { value: '75' } });
    expect(progressBar.style.width).toBe('75%');
  });
});