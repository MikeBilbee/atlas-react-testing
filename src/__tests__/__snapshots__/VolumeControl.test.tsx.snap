// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VolumeControl component > renders correctly with initial volume 1`] = `
<div>
  <div
    class="flex items-center space-x-2"
  >
    <svg
      aria-hidden="true"
      class="h-5 w-5 text-gray-600 dark:text-gray-400"
      data-slot="icon"
      data-testid="speaker-icon"
      fill="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M13.5 4.06c0-1.336-1.616-2.005-2.56-1.06l-4.5 4.5H4.508c-1.141 0-2.318.664-2.66 1.905A9.76 9.76 0 0 0 1.5 12c0 .898.121 1.768.35 2.595.341 1.24 1.518 1.905 2.659 1.905h1.93l4.5 4.5c.945.945 2.561.276 2.561-1.06V4.06ZM18.584 5.106a.75.75 0 0 1 1.06 0c3.808 3.807 3.808 9.98 0 13.788a.75.75 0 0 1-1.06-1.06 8.25 8.25 0 0 0 0-11.668.75.75 0 0 1 0-1.06Z"
      />
      <path
        d="M15.932 7.757a.75.75 0 0 1 1.061 0 6 6 0 0 1 0 8.486.75.75 0 0 1-1.06-1.061 4.5 4.5 0 0 0 0-6.364.75.75 0 0 1 0-1.06Z"
      />
    </svg>
    <div
      class="relative min-w-64 sm:w-35 md:w-45 lg:w-96 h-2 bg-gray-200 dark:bg-gray-400 rounded-full"
    >
      <div
        class="absolute h-full bg-light-background dark:bg-dark-secondary rounded-full transition-all duration-300 ease-in-out"
        style="width: 50%;"
      />
      <input
        class="
						absolute w-full h-full appearance-none bg-transparent focus:outline-none
						&::-webkit-slider-thumb {
							-webkit-appearance: none;
							appearance: none;
							width: 12px;
							height: 12px;
							background-color: #4b5563;
							border-radius: 50%;
							cursor: pointer;
						}
					"
        max="100"
        min="0"
        type="range"
        value="50"
      />
    </div>
  </div>
</div>
`;

exports[`VolumeControl component > renders with initial volume and speaker icon 1`] = `
<div>
  <div
    class="flex items-center space-x-2"
  >
    <svg
      aria-hidden="true"
      class="h-5 w-5 text-gray-600 dark:text-gray-400"
      data-slot="icon"
      fill="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M13.5 4.06c0-1.336-1.616-2.005-2.56-1.06l-4.5 4.5H4.508c-1.141 0-2.318.664-2.66 1.905A9.76 9.76 0 0 0 1.5 12c0 .898.121 1.768.35 2.595.341 1.24 1.518 1.905 2.659 1.905h1.93l4.5 4.5c.945.945 2.561.276 2.561-1.06V4.06ZM18.584 5.106a.75.75 0 0 1 1.06 0c3.808 3.807 3.808 9.98 0 13.788a.75.75 0 0 1-1.06-1.06 8.25 8.25 0 0 0 0-11.668.75.75 0 0 1 0-1.06Z"
      />
      <path
        d="M15.932 7.757a.75.75 0 0 1 1.061 0 6 6 0 0 1 0 8.486.75.75 0 0 1-1.06-1.061 4.5 4.5 0 0 0 0-6.364.75.75 0 0 1 0-1.06Z"
      />
    </svg>
    <div
      class="relative min-w-64 sm:w-35 md:w-45 lg:w-96 h-2 bg-gray-200 dark:bg-gray-400 rounded-full"
    >
      <div
        class="absolute h-full bg-light-background dark:bg-dark-secondary rounded-full transition-all duration-300 ease-in-out"
        style="width: 50%;"
      />
      <input
        class="
						absolute w-full h-full appearance-none bg-transparent focus:outline-none
						&::-webkit-slider-thumb {
							-webkit-appearance: none;
							appearance: none;
							width: 12px;
							height: 12px;
							background-color: #4b5563;
							border-radius: 50%;
							cursor: pointer;
						}
					"
        max="100"
        min="0"
        type="range"
        value="50"
      />
    </div>
  </div>
</div>
`;
